utils/error_injection.hh

```diff
@@ -135,6 +207,14 @@ class error_injection {
         return it->second.one_shot;
     }
 
+    std::optional<std::reference_wrapper<injection_data>> get_data(const std::string_view& injection_name) {
```

Comment:
There's a simpler way to express `std::optional<std::reference_wrapper<injection_data>>`.
It's called `injection_data*`.

Reply:
Pointer points to some place in the memory, it doesn't tell us anything about the lifetime of the object it points to. But when I see a reference, I know that this object is alive at least as long as this reference is alive.

Reply:
> Pointer points to some place in the memory, it doesn't tell us anything about the lifetime of the object it points to. But when I see a reference, I know that this object is alive at least as long as this reference is alive.

It tells you the same things about lifetime as a reference does -- at least that's the rule you should follow when writing modern C++.

- `unique_ptr` - unique ownership, the object dies when the pointer dies
- `shared_ptr` - shared ownership, the object dies when the pointer dies if we were the last owner
- raw pointer - no ownership, the object outlives the pointer

Raw pointers in modern C++ act as references, the only difference being that they can be `nullptr`.

So `injection_data*` is just like a reference which can be null. I.e. a simpler way to express `std::optional<std::reference_wrapper<injection_data>>`

For now I'm gonna leave you the choice if you prefer `std::optional<std::reference_wrapper<injection_data>>` because I see that we use this monster in a few places in our code. (But we also use raw pointers for the same purpose.)

Maybe in the future we'll write some commandments for such things in the form of a doc stored in the repo...

Reply:
I didn't know what's the preferred way, so I chose the one I thought is good and I found in scylla code. But if
> raw pointer - no ownership, the object outlives the pointer

I will change that to a raw pointer.